### PR TITLE
touch panel to display adjustment

### PR DESCRIPTION
**Pull request by holzingerm**
implement the ability to align touch panel coordinate system to display
coordinate system if the are not matching by default.
All settings will be implemented in .kodi/userdata/advancedsettings.xml
if they are required otherwise default behaviour is used

### CODE REVIEW

**Comment by holzingerm** [comment]
I don't think so.

If the touch panels origin is > than 1080 than you are not able to perfectly align it to the origin of the display. best would be if I find a simple solution to implement the range dynamically to the screen resolution. (max of with and high)

for the moment it is suitable up to a screen resolution of 1920x1920 (standard resolution FullHD = 1920x1080)

BTW: on my setup i only need the stretch factors but many users reported to have also offset differences.

see
http://4.bp.blogspot.com/-CcNJvQzcnLg/USp0Cl9FGsI/AAAAAAAAAL8/UKIsLP169NI/s1600/sch.png
for a explanation of the codes intended functionality

**Comment by garbear** [comment]
advanced settings are only loaded once on startup, but the resolution can change at any time, so doing this dynamically won't work

What ranges can x and y offset take? Why limit them to positive values? Maybe changing both limits to the range [-3840, 3840] will be the most flexible

**Comment by holzingerm** [comment]
Correct.

Will delete the limits to allow the whole integer range to be used.

File: xbmc/settings/AdvancedSettings.cpp [hunk 1]
```diff
@@ -888,6 +894,16 @@ void CAdvancedSettings::ParseSettingsFile(const std::string &file)
     XMLUtils::GetInt(pElement, "commbreakautowind", m_iEdlCommBreakAutowind, 0, 10);        // Between 0 and 10 seconds
   }
 
+  // Touchscreencontroller allignment
```
Comment (mkortstiege) [inline xbmc/settings/AdvancedSettings.cpp]: `Touchscreen alignment` please

File: xbmc/settings/AdvancedSettings.cpp [hunk 2]
```diff
@@ -888,6 +894,16 @@ void CAdvancedSettings::ParseSettingsFile(const std::string &file)
...
+  pElement = pRootElement->FirstChildElement("touchscreen_allign");
```
Comment (mkortstiege) [inline xbmc/settings/AdvancedSettings.cpp]: I'd just name it `touchscreen`

File: xbmc/settings/AdvancedSettings.h [hunk 3]
```diff
@@ -344,6 +344,12 @@ class CAdvancedSettings : public ISettingCallback, public ISettingsHandler
     std::string m_cpuTempCmd;
     std::string m_gpuTempCmd;
 
+    // Touchscreen allignment advanced settings
```
Comment (mkortstiege) [inline xbmc/settings/AdvancedSettings.h]: just one l ;)

**Comment by holzingerm** [comment]
Create a new commit to correct my typos/C&P mistakes and rename the comments and entries on all files (2) to "touchscreencontroller"

**Comment by fritsch** [comment]
Please force push next time - no need to close and reopen new PRs.

**Comment by mkortstiege** [comment]
Please squash. Btw, I still think `touchscreen`is enough as it reads and types easier.

**Comment by holzingerm** [comment]
thanks for the "lessons learned" (squash commits)

- all commits are "squashed" to the initial one
- renamed the advancedsettings.xml section to "touchscreen"

**Comment by mkortstiege** [comment]
Thanks. @koying mind having a look?

**Comment by holzingerm** [comment]
Hi

Is there anything I can do to make this commit be reviewed, or is it only a matter of time when the commit will be merged ore rejected.

**Comment by koying** [comment]
I don't quite understand the issue this PR solves, really.
Which device exhibit this shift of mouse coordinates?

**Comment by holzingerm** [comment]
The issue is on all touchscreen devices were the display resolution is not aligned to the touchscreen coordinates.
This could be mechanically be solved, per touchscreen firmware or per software.

Let me be a little bit more specific:
If the touchscreen resolution is for example 1280x800 the touch input should place the touch/mouse point exactly within this range of points on the  screen when pressing on the screen.

if the touchscreen controller sends its touch position in his own coordination system for example 1920x1920  we need to align this system to the screen.

issue on many self made media centers:
with standard kodi compiled from source or instaleld by package manager (or OpenElec,...) the mouse do not follow the touch input exactly.
in my case I'm only able to use 2/3 of the screen per touch input 
==> when my finger is a the end of the touchscreen on the very bottom right, the mouse is only at the 2/3 of the display
So I can not get to the bottom right corner and need a additional mouse to handle elements in that dead section.

**Comment by holzingerm** [comment]
The image will also show the behavior I'd like to address
http://4.bp.blogspot.com/-CcNJvQzcnLg/USp0Cl9FGsI/AAAAAAAAAL8/UKIsLP169NI/s1600/sch.png

it came all out of the CarPC project here where I'd like to support
http://www.engineering-diy.blogspot.ro/

**Comment by holzingerm** [comment]
Thanks so far for reviewing my first commit to an official repository

best regards
Mario Holzinger

**Comment by koying** [comment]
> If the touchscreen resolution is for example 1280x800 the touch input should place the touch/mouse point exactly within this range of points on the screen when pressing on the screen.
> 
> if the touchscreen controller sends its touch position in his own coordination system for example 1920x1920 we need to align this system to the screen.

Ok, but that's a firmware issue, do we agree?
While I understand your need for this workaround, very few kodi users would benefit from it. 

My opinion is to not impact a vast majority of users with additional CPU cycles for this edge case, so -1 from me to incorporate in master.

Thanks for submitting, anyway. If someone else faces the same problem, they'll be happy to find this.

**Comment by holzingerm** [comment]
the strange behavior is that on X-window the issue do not occure. only kodi so far. maybe because kodi is fed directly from the device and X-window allready make some alignment.

thx

**Comment by vpeter4** [comment]
@koying: Not every touchscreen controller sends absolute coordinates. Some sends 0,0 for top,left and max_x,max_y for bottom,right. Others can send 0,0 for top,left and some very large data for bottom,right. And some other doesn't send either 0,0 for top,left but instead some very large data. That's why some calculations must be done.
Usually kernel module handles touchscreen and we would expect this function there. But kernel devs doesn't agree. That's why it must be done in userspace.
This is handled in X server. But in case Kodi runs without X this must be handled in Kodi itself. Or with some other userspace program.
For my personal use in OpenELEC I made one daemon which uses tslib touchscreen library and creates one new virtual input device Kodi uses. Also calibration addon to calibrate in case touchscreen needs that. But if Kodi is getting some touchscreen support I would expect this function also to be present.

**Comment by mkortstiege** [comment]
No objection from my side as this is not really adding anything that might cause breakage.
@Memphiz, @Montellese and @koying what do you think about this?

**Comment by holzingerm** [comment]
Due to the fact that the modified file have changed till now shall I solve the merg conflicts and squash the commit again to make merging easier if it would be accepted by the pro's?

**Comment by holzingerm** [comment]
may I also provide the team with an Wiki text example to save time on your side

**Comment by garbear** [comment]
documentation? what's that :p sorry we didn't get this in for jarvis. we're in a feature freeze right now so I'll shove this in when we branch for 17 alpha 1 next month

**Comment by NedScott** [comment]
@holzingerm Anything you can provide for wiki/documentation would be great. Even if this hasn't been pulled in yet, we can get documentation ready and just note that it isn't in master yet. You can either request an account on the wiki and make a new page, or just poke me and tell me what you want it to say. Either is fine.

**Comment by holzingerm** [comment]
sucessfully rebased to xbmc/master

[advancedsettings.txt](https://github.com/xbmc/xbmc/files/59581/advancedsettings.txt)

**Comment by garbear** [comment]
@holzingerm I updated the wiki: http://kodi.wiki/view/Advancedsettings.xml#touchscreen

does that look alright to you?

**Comment by holzingerm** [comment]
Yes that looks good. :-)
thank you for that.

**Comment by garbear** [comment]
Cool. looks like this can go in now

**Comment by maxnet** [comment]
>sucessfully rebased to xbmc/master

Have you also tested it with a multi-touch screen since, or only with one that emulates single mouse events?

Your code only alters the coordinates in AbsEvent()
However input from multi-touch devices is now handled in mtAbsEvent() since PR #8129 was merged.

**Comment by popcornmix** [comment]
@maxnet does (untested) https://github.com/popcornmix/xbmc/commit/5cc885f5da94e07159855788e6f3d9a1fb4acf2c look okay?

**Comment by maxnet** [comment]
>@maxnet does (untested) popcornmix@5cc885f look okay?

Looks ok.
But didn't test it either.

**Comment by xelus** [comment]
Tested it on an usbest technology sis hid touch controller. Doesn't work there - see http://forum.kodi.tv/showthread.php?tid=259500.

**Pull request merged**